src/app/services/xml-parsers/witnesses-parser.service.ts

```diff
@@ -0,0 +1,134 @@
+    return lists.filter((list) => !isNestedInElem(list, list.tagName))
+      .map((list) => this.parseWitnesses(list))
+      .reduce((x, y) => x.concat(y), [])
+      .reduce((list, item) => {
+        list[item.id] = item;
+
+        return list;
+      },      {});
+  }
+
+  private parseWitnesses(list: XMLElement) {
+    return Array.from(list.querySelectorAll<XMLElement>(this.witTagName))
+      .map((wit) => this.parseWitness(wit));
+  }
+
+  private parseWitness(wit: XMLElement): Witness {
+    return {
+      id: wit.getAttribute('xml:id') || xpath(wit),
+      name: this.parseWitnessName(wit) || wit.getAttribute('xml:id') || xpath(wit),
+      attributes: this.genericParserService.parseAttributes(wit),
+      content: this.parseWitnessContent(wit),
+      group: this.parseParentGroupId(wit),
+    };
+  }
+
+  private parseWitnessName(wit: XMLElement): Description {
+    const witNameEl = wit.querySelector<XMLElement>('*[' + this.witNameAttr + ']');
+    let witName: Description;
+
+    if (witNameEl) {
+      witName = Array.from(witNameEl.childNodes)
+        .map((child: XMLElement) => this.genericParserService.parse(child));
+    }
+
+    return witName;
+  }
+
+  private parseWitnessContent(wit: XMLElement): Description {
+    return Array.from(wit.childNodes)
+      .filter((child) => child.nodeName !== this.witListTagName && child.textContent.trim().length !== 0)
+      .map((child: XMLElement) => this.genericParserService.parse(child));
+  }
+
+  private parseWitnessesGroups(lists: XMLElement[]): WitnessGroups {
+    return {
+      groups: this.parseGroups(lists),
+    };
+  }
+
+  private parseGroups(lists: XMLElement[]) {
+    return lists.filter((list) => isNestedInElem(list, list.tagName))
+      .map((list) => this.parseWitnessGroup(list))
+      .reduce((x, y) => x.concat(y), [])
+      .reduce((list, item) => {
+        list[item.id] = item;
+
+        return list;
+      },      {});
+  }
+
+  private parseWitnessGroup(list: XMLElement): WitnessGroup {
+    return {
+      id: list.getAttribute('xml:id') || xpath(list),
+      name: this.parseGroupName(list) || replaceNotWordChar(list.getAttribute('xml:id')) || xpath(list),
+      attributes: this.genericParserService.parseAttributes(list),
+      content: this.parseGroupContent(list),
+      group: this.parseParentGroupId(list),
+    };
+  }
+
+  private parseGroupName(list: XMLElement) {
+    const groupTagName = 'head';
+    const groupEl = Array.from(list.children)
+      .find((child) => child.nodeName === groupTagName);
+
+    return groupEl && groupEl.textContent;
+  }
+
+  private parseGroupContent(list: XMLElement) {
+    return Array.from(list.children)
+      .filter(({nodeName}) => nodeName === this.witListTagName || nodeName === this.witTagName)
+      .map((child) => child.getAttribute('xml:id'));
+  }
+
+  private parseParentGroupId(element: XMLElement) {
+    let parentEl = element.parentElement;
+
+    if (parentEl) {
```

Comment:
What happens is `parentEl` is `undefined`?

Reply:
This line is unnecessary and can be removed. parentEl will never be undefined.